### PR TITLE
Configuration: You can now see your expired API keys if you have no active ones

### LINKED ISSUE #42404: Unable to see deleted API keys if there are no valid API keys
**What happened**:
If all the API keys have expired, then from the Grafana API keys page, there is no way to get to the list of expired API keys.

There are two API keys in the database
```
sqlite> select * from api_key;
1|1|admin|9f93564bd3487fd6a9030e26f358bcd1b75b2be81a0bd9415e125277a0a418cab7baeb413ae1c6483e37121fdcde6af5ece3|Admin|2021-07-22 15:14:54|2021-07-22 15:14:54|1627053294
2|1|admin1|6f32b5a25301a249e67256083ee095470ac571006d43c391003a61c2d43f5c6c76c495bb0fb8f501c9b75e4de9ebd2453d7f|Admin|2021-11-15 13:22:59|2021-11-15 13:22:59|1637068979
```
but non visible on the API key page, until you create one. 

hidden_apikeys

**What you expected to happen**:
If there are expired API keys, it should be possible to view and delete them. 

**How to reproduce it (as minimally and precisely as possible)**:
Start with a clean database. 
create an apikey with name `viewer1` and a 1s expiry time 
go back to the API key page. 
try to create another API key with the name `viewer1`

**Anything else we need to know?**:
Only a minor annoyance really. Once you twig what the problem is. 

**Environment**:
- Grafana version: 8.2.5 
- Data source type & version: sqlite
- OS Grafana is installed on: wsl2 
- User OS & Browser: windows /firefox.
- Grafana plugins: n/a
- Others:

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

- on initial load of the configuration page it currently doesn't query for expired keys. this meant that if there were some expired keys but no active keys it would show the call to action banner, so there was then no way to toggle showing expired keys.
- instead, always fetch all keys. this matches the behaviour of something like github. it should also hopefully encourage people to clean up their expired keys.
- add new optional `isExpired` value to api return that makes it easier to style the frontend. this _could_ be calculated on the frontend using the expiration time? but felt like we could get into a mess with timezones there 🤷 
- apply some minor styling/tooltip to expired api keys

| Before | After |
| --- | --- |
| image | image |

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/42404

**Special notes for your reviewer**:

- another option could be to put expired keys into a separate section?
- i'm not 100% sure why we had expired keys behind a toggle. my guess would be performance. but i think we should be nudging people to delete expired keys... 🤔

### CODE REVIEW

File: public/app/features/api-keys/ApiKeysTable.tsx [hunk 1]
```diff
@@ -25,10 +29,19 @@ export const ApiKeysTable: FC<Props> = ({ apiKeys, timeZone, onDelete }) => {
         <tbody>
           {apiKeys.map((key) => {
```
Comment (axelavargas) [inline public/app/features/api-keys/ApiKeysTable.tsx:30]: ❓ do you know if the API keys are sorted using time? I am thinking that removing the toggle can lead to the display of a ton of expired keys together with valid ones, making it a bit hard to see which ones are valid, also cleaning them is going to be quite painful (clicking delete one by one)

**Review by axelavargas — COMMENTED** [review]
Hola, Hola 👋🏾 @ashharrison90, great job in this PR 👏🏾 , I like we have introduced a  visual differentiation between expired keys and valid ones. I left some comments regarding usability :)

**Review by hugohaggmark — COMMENTED** [review]
I'm not sure we should be defaulting to always passing `includeExpired` as `true`. For larger installations that could cause havoc. We should at least limit the results to 100 and introduce paging so we don't cause any issues.

File: public/app/features/api-keys/ApiKeysTable.tsx (hunk 1 above)
Comment (ashharrison90) [reply public/app/features/api-keys/ApiKeysTable.tsx:30]: i believe they're sorted alphabetically currently

**Comment by ashharrison90** [comment]
@hugohaggmark @axelavargas did some digging on this, turns out this call is always limited to the first 100 alphabetically here: https://github.com/grafana/grafana/blob/ash/42404/pkg/services/sqlstore/apikey.go#L27



i can think of a couple of options (as well as any you can think of as well!):
- always retrieve expired keys (as in this PR).
  - upsides: simple
  - downsides: keys mixed together, if there are more than 100 keys some won't be shown. we have this problem already but it's definitely slightly worse by including expired keys as well. 
- keep the toggle, but always make both calls and put them in the store. this way we can check for the existence of expired keys if there are no active keys.
  - upsides: toggle remains!
  - downsides: makes the store a bit more complex (2 slightly different lists of apiKeys), doubles the amount of calls to the backend (even if they're lightweight)
- keep the toggle, move it's state to the store, and only check for expired keys if the toggle is off and there are no keys.
  - upsides: toggle remains, store remains simpleish (still only 1 list of apiKeys)
  - downsides: introduces some extra logic to our actions which could be an antipattern. i've seen discussions both ways on this tbh. could also be slightly confusing, e.g. if you have expired keys but no active keys the toggle would default to true. if you have expired keys and delete your last active key the toggle state would change to true.

i'm leaning towards the first option as it's definitely simpler and performance is no longer a concern, but curious to know your thoughts 🙏

**Comment by axelavargas** [comment]
> @hugohaggmark @axelavargas did some digging on this, turns out this call is always limited to the first 100 alphabetically here: [`ash`/42404/pkg/services/sqlstore/apikey.go#L27](https://github.com/grafana/grafana/blob/ash/42404/pkg/services/sqlstore/apikey.go?rgh-link-date=2021-12-01T17%3A28%3A16Z#L27)
> 
> i can think of a couple of options (as well as any you can think of as well!):
> 
> * always retrieve expired keys (as in this PR).
>   
>   * upsides: simple
>   * downsides: keys mixed together, if there are more than 100 keys some won't be shown. we have this problem already but it's definitely slightly worse by including expired keys as well.
> * keep the toggle, but always make both calls and put them in the store. this way we can check for the existence of expired keys if there are no active keys.
>   
>   * upsides: toggle remains!
>   * downsides: makes the store a bit more complex (2 slightly different lists of apiKeys), doubles the amount of calls to the backend (even if they're lightweight)
> * keep the toggle, move it's state to the store, and only check for expired keys if the toggle is off and there are no keys.
>   
>   * upsides: toggle remains, store remains simpleish (still only 1 list of apiKeys)
>   * downsides: introduces some extra logic to our actions which could be an antipattern. i've seen discussions both ways on this tbh. could also be slightly confusing, e.g. if you have expired keys but no active keys the toggle would default to true. if you have expired keys and delete your last active key the toggle state would change to true.
> 
> I'm leaning towards the first option as it's definitely simpler and performance is no longer a concern, but curious to know your thoughts pray

Hey @ashharrison90, great job investigating this a bit further 🙌🏾 , my thinking is leaning towards option 2, even with the hit in performance 😓 , I think option 1 can have a clear impact on usability, It will be quite hard to identify which API keys are active if we could have 100 mixed API keys (without any filtering mechanism), as well I am wondering: 

If the ApiKeys are sorted alphabetically and capped to 100, does it mean in option 1 we could have a scenario when we retrieve 100 expired keys? 

Last minor thing, I wanted to test  the PR and found a whenever I go to the API-Keys it was giving me an error (infinite loading), I am not sure if is only happening to me:

<details>
<summary>Dev Tools error </summary>

{
  "error": "runtime error: invalid memory address or nil pointer dereference - /usr/local/go/src/runtime/panic.go:221 (0x454a66)\n\tpanicmem: panic(memoryError)\n/usr/local/go/src/runtime/signal_unix.go:735 (0x454a36)\n\tsigpanic: panicmem()\n/home/axelav/proyects/grafana/pkg/api/apikey.go:38 (0x2165cef)\n\tGetAPIKeys: result[i].IsExpired = time.Now().After(*expiration)\n/home/axelav/proyects/grafana/pkg/api/routing/routing.go:17 (0x169dc41)\n\tWrap.func1: if res := handler(c); res != nil {\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0x1699b73)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/middleware/request_metrics.go:55 (0x1699b62)\n\tRequestMetrics.func1.1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0x23c47a4)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/extensions/caching/middleware/query_middleware.go:487 (0x23c4794)\n\t(*QueryMiddleware).Handle: c.Next()\n/home/axelav/proyects/grafana/pkg/macaron/context.go:32 (0xdbebbb)\n\tContextInvoker.Invoke: invoke(params[0].(*Context))\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:139 (0xdbfb54)\n\t(*injector).fastInvoke: return f.Invoke(in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:117 (0xdbfa7c)\n\t(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0xdc6dba)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/macaron/macaron.go:155 (0xdc6dac)\n\t(*Macaron).UseMiddleware.func1: c.Next()\n/usr/local/go/src/net/http/server.go:2046 (0x6db46e)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/home/axelav/proyects/grafana/pkg/middleware/csp.go:22 (0x1696372)\n\tAddCSPHeader.func1.1: next.ServeHTTP(rw, req)\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0xdc6dba)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/macaron/macaron.go:155 (0xdc6dac)\n\t(*Macaron).UseMiddleware.func1: c.Next()\n/usr/local/go/src/net/http/server.go:2046 (0x6db46e)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/home/axelav/proyects/grafana/pkg/macaron/render.go:44 (0xdc1aa6)\n\tRenderer.func1.1: next.ServeHTTP(rw, req)\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0x1698eca)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/middleware/recovery.go:167 (0x1698ec1)\n\tRecovery.func1: c.Next()\n/home/axelav/proyects/grafana/pkg/macaron/context.go:32 (0xdbebbb)\n\tContextInvoker.Invoke: invoke(params[0].(*Context))\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:139 (0xdbfb54)\n\t(*injector).fastInvoke: return f.Invoke(in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:117 (0xdbfa7c)\n\t(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0x1696d84)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/middleware/logger.go:33 (0x1696d6d)\n\tLogger.func1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/context.go:62 (0x169b759)\n\t(*Context).Next: ctx.run()\n/home/axelav/proyects/grafana/pkg/middleware/request_tracing.go:57 (0x169b748)\n\tRequestTracing.func1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x4e61b3)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), \u0026regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x4e5764)\n\tValue.Call: return v.call(\"Call\", in)\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:159 (0xdbfde6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/home/axelav/proyects/grafana/pkg/macaron/inject.go:119 (0xdbfa34)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/home/axelav/proyects/grafana/pkg/macaron/context.go:67 (0xdbeccb)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/home/axelav/proyects/grafana/pkg/macaron/router.go:156 (0xdc33dd)\n\t(*Router).Handle.func1: c.run()\n/home/axelav/proyects/grafana/pkg/macaron/router.go:223 (0xdc4071)\n\t(*Router).ServeHTTP: h(rw, req, p)\n/home/axelav/proyects/grafana/pkg/macaron/macaron.go:189 (0xdc13f0)\n\t(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)\n/usr/local/go/src/net/http/server.go:2878 (0x6de9da)\n\tserverHandler.ServeHTTP: handler.ServeHTTP(rw, req)\n/usr/local/go/src/net/http/server.go:1929 (0x6da547)\n\t(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)\n/usr/local/go/src/runtime/asm_amd64.s:1581 (0x471ea0)\n\tgoexit: BYTE\t$0x90\t// NOP\n",
  "message": "Internal Server Error - Check the Grafana server logs for the detailed error message."
}


</details>

**Comment by hugohaggmark** [comment]
> @hugohaggmark @axelavargas did some digging on this, turns out this call is always limited to the first 100 alphabetically here: https://github.com/grafana/grafana/blob/ash/42404/pkg/services/sqlstore/apikey.go#L27
> 
> i can think of a couple of options (as well as any you can think of as well!):
> 
> * always retrieve expired keys (as in this PR).
>   
>   * upsides: simple
>   * downsides: keys mixed together, if there are more than 100 keys some won't be shown. we have this problem already but it's definitely slightly worse by including expired keys as well.
> * keep the toggle, but always make both calls and put them in the store. this way we can check for the existence of expired keys if there are no active keys.
>   
>   * upsides: toggle remains!
>   * downsides: makes the store a bit more complex (2 slightly different lists of apiKeys), doubles the amount of calls to the backend (even if they're lightweight)
> * keep the toggle, move it's state to the store, and only check for expired keys if the toggle is off and there are no keys.
>   
>   * upsides: toggle remains, store remains simpleish (still only 1 list of apiKeys)
>   * downsides: introduces some extra logic to our actions which could be an antipattern. i've seen discussions both ways on this tbh. could also be slightly confusing, e.g. if you have expired keys but no active keys the toggle would default to true. if you have expired keys and delete your last active key the toggle state would change to true.
> 
> i'm leaning towards the first option as it's definitely simpler and performance is no longer a concern, but curious to know your thoughts 🙏

Great digging @ashharrison90. But wouldn't the original issue be fixed if when we load the page we always include the expired keys and then add logic for that? I must be misunderstanding the complexity for this task. Maybe we should get in a call together?

**Comment by hugohaggmark** [comment]
After a brief call with @ashharrison90 I also agree that option 2 seems like the best way forward.

**Comment by ashharrison90** [comment]
ok @axelavargas @hugohaggmark i've updated the PR to match option 2. the behaviour should be something like this:
- if there are >0 active keys and >0 expired keys, everything basically behaves as before. `Include expired keys` is false by default, but you can toggle it on if you wish.
- if there are 0 active keys but >0 expired keys, set `Include expired keys` to true and disable interacting with the toggle until there are >0 active keys. this prevents some undesirable behaviour where you could uncheck this toggle and immediately get the CTA.
- if there are 0 active keys and 0 expired keys, show the CTA

@axelavargas re: the go loop issue, i've never seen that. did you restart your grafana server? 🤔  i was worried it might be something to do with old keys that are created on main and somehow don't get the new property included... but have just tested that and it works fine.

**Comment by hugohaggmark** [comment]
Taking it out for a spin now

**Comment by hugohaggmark** [comment]
Hmm I also get an error on the backend when I add a key without entering an expiration date:

```bash
"runtime error: invalid memory address or nil pointer dereference" stack="/usr/local/go/src/runtime/panic.go:221 (0x404e626)\n\tpanicmem: panic(memoryError)\n/usr/local/go/src/runtime/signal_unix.go:735 (0x404e5f6)\n\tsigpanic: panicmem()\n/Users/hugo/git/grafana/grafana/pkg/api/apikey.go:38 (0x5d07ccf)\n\tGetAPIKeys: result[i].IsExpired = time.Now().After(*expiration)\n/Users/hugo/git/grafana/grafana/pkg/api/routing/routing.go:17 (0x5240261)\n\tWrap.func1: if res := handler(c); res != nil {\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x523c193)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/middleware/request_metrics.go:55 (0x523c182)\n\tRequestMetrics.func1.1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x49bbc9a)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/macaron/macaron.go:155 (0x49bbc8c)\n\t(*Macaron).UseMiddleware.func1: c.Next()\n/usr/local/go/src/net/http/server.go:2046 (0x42d0aee)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/hugo/git/grafana/grafana/pkg/middleware/csp.go:22 (0x5238992)\n\tAddCSPHeader.func1.1: next.ServeHTTP(rw, req)\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x49bbc9a)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/macaron/macaron.go:155 (0x49bbc8c)\n\t(*Macaron).UseMiddleware.func1: c.Next()\n/usr/local/go/src/net/http/server.go:2046 (0x42d0aee)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/hugo/git/grafana/grafana/pkg/macaron/render.go:44 (0x49b6986)\n\tRenderer.func1.1: next.ServeHTTP(rw, req)\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x523b4ea)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/middleware/recovery.go:167 (0x523b4e1)\n\tRecovery.func1: c.Next()\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:32 (0x49b3a9b)\n\tContextInvoker.Invoke: invoke(params[0].(*Context))\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:139 (0x49b4a34)\n\t(*injector).fastInvoke: return f.Invoke(in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:117 (0x49b495c)\n\t(*injector).Invoke: return inj.fastInvoke(v, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x52393a4)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/middleware/logger.go:33 (0x523938d)\n\tLogger.func1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:62 (0x523dd79)\n\t(*Context).Next: ctx.run()\n/Users/hugo/git/grafana/grafana/pkg/middleware/request_tracing.go:57 (0x523dd68)\n\tRequestTracing.func1: c.Next()\n/usr/local/go/src/reflect/value.go:543 (0x40da573)\n\tValue.call: call(frametype, fn, stackArgs, uint32(frametype.size), uint32(abi.retOffset), uint32(frameSize), &regArgs)\n/usr/local/go/src/reflect/value.go:339 (0x40d9b24)\n\tValue.Call: return v.call(\"Call\", in)\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:159 (0x49b4cc6)\n\t(*injector).callInvoke: return reflect.ValueOf(f).Call(in), nil\n/Users/hugo/git/grafana/grafana/pkg/macaron/inject.go:119 (0x49b4914)\n\t(*injector).Invoke: return inj.callInvoke(f, t, t.NumIn())\n/Users/hugo/git/grafana/grafana/pkg/macaron/context.go:67 (0x49b3bab)\n\t(*Context).run: if _, err := ctx.Invoke(ctx.handler()); err != nil {\n/Users/hugo/git/grafana/grafana/pkg/macaron/router.go:156 (0x49b82bd)\n\t(*Router).Handle.func1: c.run()\n/Users/hugo/git/grafana/grafana/pkg/macaron/router.go:223 (0x49b8f51)\n\t(*Router).ServeHTTP: h(rw, req, p)\n/Users/hugo/git/grafana/grafana/pkg/macaron/macaron.go:189 (0x49b62d0)\n\t(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)\n/usr/local/go/src/net/http/server.go:2878 (0x42d405a)\n\tserverHandler.ServeHTTP: handler.ServeHTTP(rw, req)\n/usr/local/go/src/net/http/server.go:1929 (0x42cfbc7)\n\t(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)\n/usr/local/go/src/runtime/asm_amd64.s:1581 (0x406bd00)\n\tgoexit: BYTE\t$0x90\t// NOP\n"
```

image

**Comment by ashharrison90** [comment]
@hugohaggmark ahhhhhhhh that's why, i haven't handled the case where the expiry time isn't specified 🤦  thanks both! 🙌

**Comment by hugohaggmark** [comment]
> @hugohaggmark ahhhhhhhh that's why, i haven't handled the case where the expiry time isn't specified 🤦 thanks both! 🙌

No worries, be sure to add tests (if possible) in the backend too so we can catch those nasty things early.

**Comment by ashharrison90** [comment]
should be fixed 👍 

re: tests, there currently aren't any for `apikey.go` that i can see. i'm not sure my go is quite up to creating a whole new test suite 😬

**Comment by hugohaggmark** [comment]
> should be fixed 👍
> 
> re: tests, there currently aren't any for `apikey.go` that i can see. i'm not sure my go is quite up to creating a whole new test suite 😬

That's fair enough. Maybe @sdboyer or @ying-jeanne could add a review and leave feedback on how we should tackle this?

**Comment by sdboyer** [comment]
My largely-ignorant-of-this-system view is that it looks like we're not making things terribly worse than they already are to add this without some Go tests, but i'm gonna defer to @zserge for a more qualified opinion 😄

**Review by axelavargas — APPROVED** [review]
Fantastic job Mr. Harrison 🚀 , I gave it a spin, and works as expected!

**Comment by ashharrison90** [comment]
@zserge if you get a chance next week would appreciate a quick backend review on this 🙏  thanks! 😄

**Comment by ashharrison90** [comment]
have done the check in the frontend instead to avoid changing the backend api 👍

**Review by joshhunt — APPROVED** [review]

**Pull request merged**